tensorpipe/core/buffer_helpers.h

```diff
@@ -11,30 +11,24 @@
 #include <tensorpipe/config.h>
 #include <tensorpipe/core/buffer.h>
 
-#define TP_BUFFER_CPU_FIELD_AND_ACCESSOR(type) \
-  type cpu;                                    \
-  template <>                                  \
-  constexpr auto& get<CpuBuffer>() {           \
-    return cpu;                                \
-  }
-
 #if TENSORPIPE_SUPPORTS_CUDA
-#define TP_BUFFER_CUDA_FIELD_AND_ACCESSOR(type) \
-  type cuda;                                    \
-  template <>                                   \
-  constexpr auto& get<CudaBuffer>() {           \
-    return cuda;                                \
-  }
+#define TP_IF_CUDA(x) x
 #else
-#define TP_BUFFER_CUDA_FIELD_AND_ACCESSOR(type)
-#endif
+#define TP_IF_CUDA(x)
+#endif // TENSORPIPE_SUPPORTS_CUDA
 
-#define TP_BUFFER_FIELD_AND_ACCESSOR(cpu_type, cuda_type) \
-  struct {                                                \
-    template <typename TBuffer>                           \
-    constexpr auto& get();                                \
-    TP_BUFFER_CPU_FIELD_AND_ACCESSOR(cpu_type);           \
-    TP_BUFFER_CUDA_FIELD_AND_ACCESSOR(cuda_type);         \
+#define TP_DEVICE_FIELD(cpu_type, cuda_type)                               \
+  struct {                                                                 \
+    template <typename TBuffer>                                            \
+    constexpr auto& get() {                                                \
+      if (std::is_same<TBuffer, CpuBuffer>::value) {                       \
+        return cpu;                                                        \
+      }                                                                    \
+      TP_IF_CUDA(                                                          \
+          if (std::is_same<TBuffer, CudaBuffer>::value) { return cuda; }); \
```

Comment:
Not sure if the TP_IF_CUDA is buying us much in terms of readability in this particular instance, what do you think?

Reply:
We need it here because `CudaBuffer` is not defined unless `TENSORPIPE_SUPPORTS_CUDA` is defined.

Reply:
My point was that doing the full `#if TENSORPIPE_SUPPORTS_CUDA` thing is perhaps clearer here than passing an `if (...)` expression as an argument of a function-like macro. For example, that confuses clang-format, which formats it differently than the previous if. I found it confusing, but if you like it I'm ok to stick to it.

Reply:
That would not work because we'd need an `#if` inside the macro definition.

Reply:
Oh, right, I keep forgetting that this doesn't work. Fine to keep it like this then.